capsules/src/net/sixlowpan.rs

```diff
@@ -602,166 +736,146 @@ impl<'a, A: time::Alarm> RxClient for FragState<'a, A> {
+    ///
+    /// Only one transmission is allowed at a time. Calling this method while
+    /// before a previous tranismission has completed will return an error.
+    ///
+    /// # Arguments
+    ///
+    /// * `src_mac_addr` - Why is the argument specified?
+    ///
+    /// * `dst_mac_addr` - Why is the argument specified?
+    ///
+    /// * `ip6_packet` - A buffer containing the IPv6 packet to transmit. This
+    /// buffer will be passed back in the
+    /// [send_done](trait.SixlowpanClient.html#send_done) callback.
+    ///
+    /// * `ip6_packet_len` - The length of the packet, in case the `ip6_packet`
+    /// buffer is larger than the actual packet. Must be at most
+    /// `ip6_packet.len()`
+    ///
+    /// * `security` - An optional tuple specifying the security level
+    /// (encryption, MIC, or both) and key identifier to use. If elided, no
+    /// security is used.
+    ///
+    /// This function exposes the primary packet transmission fuctionality. The
+    /// source and destination mac address arguments specify the link-layer
+    /// addresses of the packet, while the `security`, `fragment`, and `compress`
+    /// options specify the security level (if any), whether to fragment the packet
+    /// if it is too large, and whether to compress the packet respectively. These
+    /// different options are exposed to provide the caller with more control over
+    /// how this layer modifies packets; no compression and no fragmentation can
+    /// be set, and the packet will be sent to the radio as a raw IPv6 packet.
+    ///
+    /// The `ip6_packet` argument contains a pointer to a buffer containing a valid
+    /// IPv6 packet, while the `ip6_packet_len` argument specifies the number of
+    /// bytes to send. Note that `ip6_packet.len() > ip6_packet_len`, but we check
+    /// the invariant that `ip6_packet_len <= ip6_packet.len()`.
     pub fn transmit_packet(&self,
                            src_mac_addr: MacAddress,
                            dst_mac_addr: MacAddress,
                            ip6_packet: &'static mut [u8],
                            ip6_packet_len: usize,
                            security: Option<(SecurityLevel, KeyId)>,
-                           tx_state: &'a TxState<'a>,
                            fragment: bool,
                            compress: bool)
-                           -> Result<ReturnCode, ReturnCode> {
-
-        tx_state.init_transmit(src_mac_addr,
-                               dst_mac_addr,
-                               ip6_packet,
-                               ip6_packet_len,
-                               security,
-                               fragment,
-                               compress);
-        // Queue tx_state
-        self.tx_states.push_tail(tx_state);
-        if self.tx_busy.get() {
-            Ok(ReturnCode::SUCCESS)
+                           -> Result<ReturnCode, (ReturnCode, &'static mut [u8])> {
```

Comment:
Will we ever return any variant of `Ok(...)` that is not `ReturnCode::SUCCESS`? If not, it doesn't make sense to have two kinds of return codes, and should be something more like `Result<(), (ReturnCode, &'static mut [u8]>`.

Reply:
I'll have to review the code again, but no, I don't think there's ever a variant of `Ok(...)` that is not `SUCCESS`.